File: Exercise_1.cpp
/*
------------------------------------------
* Binary Search Implememtation (Recursive)
------------------------------------------

* Time Complexity: O(log(n))

* Space Complexity: O(log(n))

Did this code successfully run on Leetcode : NA
Any problem you faced while coding this : NO

*/

#include <stdio.h>

// A recursive binary search function. It returns
// location of x in given array arr[l..r] is present,
// otherwise -1
int binarySearch(int arr[], int l, int r, int x) {
    if (l > r) return -1;

    int mid = l + (r - l) / 2; // This equation avoids integer overflow

    if (arr[mid] > x) return binarySearch(arr, l, mid - 1, x);
    else if (arr[mid] < x) return binarySearch(arr, mid + 1, r, x);
    else return mid;
}

int main(void) {
    int arr[] = { 2, 3, 4, 10, 40 };
    int n = sizeof(arr) / sizeof(arr[0]);
    int x = 10;
    int result = binarySearch(arr, 0, n - 1, x);
    (result == -1) ? printf("Element is not present in array")
                   : printf("Element is present at index %d \n",
                            result);
    return 0;
}

File: Exercise_1.swift
import Cocoa


// Time Complexity - Best = O(1) | Worst = O(logn)
// Space Complexity - O(logn)

func binarySearch(arr: [Int], startIndex: Int, endIndex: Int, searchElement: Int) -> Int {
if startIndex > endIndex {
    return -1
}
    let middleElementIndex = (startIndex + endIndex)/2
    let middleElement = arr[middleElementIndex]
    var resultIndex = 0
    if middleElement == searchElement {
        resultIndex =  middleElementIndex
    } else if middleElement > searchElement {
       return binarySearch(arr: arr, startIndex: 0, endIndex: middleElementIndex - 1, searchElement: searchElement)
    } else if middleElement < searchElement{
        return binarySearch(arr: arr, startIndex: middleElementIndex + 1, endIndex: endIndex, searchElement: searchElement)
    }
    return resultIndex
}
let arr = [100, 200, 300, 310, 511, 600]

print(binarySearch(arr: arr, startIndex: 0, endIndex: arr.count - 1, searchElement: 300))
